### PR TITLE
Persist bid.ext.prebid.meta From Bidding Server Response

### PR DESCRIPTION
**Pull request by SyntaxNode**
Implements https://github.com/prebid/prebid-server/issues/1922

Persists the `bid.ext.prebid.meta` object from the bidding server response with no changes required by the bid adapter.
- If a bid adapter explicitly specifies metadata, that takes precedence.
- If metadata from the bidding server is invalid / fails validation, PBS returns an error.
- If unknown metadata is specified from the bidding server, it is ignored.

### CODE REVIEW

File: openrtb_ext/bid.go [hunk 1]
```diff
@@ -38,18 +39,20 @@ type ExtBidPrebidCacheBids struct {
...
-	PrimaryCategoryID    string   `json:"primaryCatId,omitempty"`
-	SecondaryCategoryIDs []string `json:"secondaryCatIds,omitempty"`
+	AdvertiserDomains    []string        `json:"advertiserDomains,omitempty"` // or advertiserDomain?
+	AdvertiserID         int             `json:"advertiserId,omitempty"`
+	AdvertiserName       string          `json:"advertiserName,omitempty"`
+	AgencyID             int             `json:"agencyId,omitempty"`
+	AgencyName           string          `json:"agencyName,omitempty"`
+	BrandID              int             `json:"brandId,omitempty"`
+	BrandName            string          `json:"brandName,omitempty"`
+	DemandSource         string          `json:"demandSource,omitempty"`
```
Comment (mansinahar) [inline openrtb_ext/bid.go:49]: I am okay with this change but just curious to know where are these new fields `DemandSource` and `DChain` coming from? I don't see them mentioned in the issue linked

File: exchange/exchange_test.go [hunk 2]
```diff
@@ -3571,6 +3619,14 @@ func TestMakeBidExtJSON(t *testing.T) {
 			expectedBidExt:     ``,
 			expectedErrMessage: "invalid character",
 		},
+		{
```
Comment (mansinahar) [inline exchange/exchange_test.go:3622]: The added test cases look great! Can we please add just one more test case where `prebid.meta` doesn't exist in both `ext` and the bid?

File: openrtb_ext/bid.go [hunk 3]
```diff
@@ -38,18 +39,20 @@ type ExtBidPrebidCacheBids struct {
...
-	AgencyID             int      `json:"agencyId,omitempty"`
-	AgencyName           string   `json:"agencyName,omitempty"`
-	BrandID              int      `json:"brandId,omitempty"`
-	BrandName            string   `json:"brandName,omitempty"`
-	MediaType            string   `json:"mediaType,omitempty"`
-	NetworkID            int      `json:"networkId,omitempty"`
-	NetworkName          string   `json:"networkName,omitempty"`
-	PrimaryCategoryID    string   `json:"primaryCatId,omitempty"`
-	SecondaryCategoryIDs []string `json:"secondaryCatIds,omitempty"`
+	AdvertiserDomains    []string        `json:"advertiserDomains,omitempty"` // or advertiserDomain?
```
Comment (mansinahar) [inline openrtb_ext/bid.go:42]: **Nitpick:** I know this isn't your change but could you please remove this leftover comment: `// or advertiserDomain?` ?

File: util/maputil/maputil_test.go [hunk 4]
```diff
@@ -164,3 +164,84 @@ func TestReadEmbeddedString(t *testing.T) {
...
+		{
+			description: "Level 2 - Does Not Exist At All",
+			value:       map[string]interface{}{"foo": map[string]interface{}{"bar": "exists"}},
+			keys:        []string{"doesnotexist", "doesnotexist"},
+			expected:    false,
+		},
+		{
+			description: "Keys Nil",
+			value:       map[string]interface{}{"foo": "exists"},
+			keys:        []string{},
```
Comment (mansinahar) [inline util/maputil/maputil_test.go:214]: Looks like from the test description, you meant to pass `nil` for `keys` here

**Review by mansinahar — COMMENTED** [review]
LGTM! Just some minor comments

File: openrtb_ext/bid.go (hunk 1 above)
Comment (SyntaxNode) [reply openrtb_ext/bid.go:49]: The authoritative source for those metadata fields is the Prebid.js docs: https://docs.prebid.org/dev-docs/bidder-adaptor.html

File: exchange/exchange_test.go (hunk 2 above)
Comment (SyntaxNode) [reply exchange/exchange_test.go:3622]: Sure. I added a test case where there are no `ext`s. There are also many other tests cases which do not define any `meta` fields.

File: openrtb_ext/bid.go (hunk 3 above)
Comment (SyntaxNode) [reply openrtb_ext/bid.go:42]: Removed. Still my fault though, just from a previous PR.

File: openrtb_ext/bid.go (hunk 1 above)
Comment (mansinahar) [reply openrtb_ext/bid.go:49]: I see. Thanks!

**Review by mansinahar — APPROVED** [review]

File: exchange/exchange.go [hunk 5]
```diff
@@ -979,10 +977,22 @@ func makeBidExtJSON(ext json.RawMessage, prebid *openrtb_ext.ExtBidPrebid, impEx
 		extMap = make(map[string]interface{})
 	}
 
+	// ext.prebid
+	if prebid.Meta == nil && maputil.HasElement(extMap, "prebid", "meta") {
```
Comment (VeronikaSolovei9) [inline exchange/exchange.go:981]: My idea was to reuse existing `jsonutil`, but I'm ok with this too :)

**Review by VeronikaSolovei9 — APPROVED** [review]

File: exchange/exchange.go (hunk 5 above)
Comment (VeronikaSolovei9) [reply exchange/exchange.go:981]: In `fpd` branch I modified `jsonutil` to support exactly what you wrote but with byte array:
`func FindElement(extension []byte, elementNames ...string)` - to check element. with any nexting level exists
and 
`func SetElement(originDataInput []byte, setValue []byte, keys ...string)` - to set element with any nested level.
Using this functions will replace logic related to `extMap` variable and Unmarshal and Marshal will not be necessary anymore

Comment (SyntaxNode) [reply exchange/exchange.go:981]: Thank you for sharing. The `jsontuil` methods operate on a json byte array and new `maputil` is operating on an already unmashalled map. There are other functions in this method which require full blown json marshal. So, I don't think the `jsontuil` would be a good use in this case.

**Pull request merged**